hydra-node/src/Hydra/Network/Reliability.hs

```diff
@@ -0,0 +1,294 @@
+  NetworkComponent m (Authenticated (ReliableMsg (Heartbeat msg))) (ReliableMsg (Heartbeat msg)) a ->
+  NetworkComponent m (Authenticated (Heartbeat msg)) (Heartbeat msg) a
+withReliability tracer me otherParties withRawNetwork callback action = do
+  ackCounter <- newTVarIO $ replicate (length allParties) 0
+  sentMessages <- newTVarIO IMap.empty
+  seenMessages <- newTVarIO $ Map.fromList $ (,0) <$> toList allParties
+  resendQ <- newTQueueIO
+  ourIndex <- findPartyIndex me
+  let resend = writeTQueue resendQ
+  withRawNetwork (reliableCallback ackCounter sentMessages seenMessages resend) $ \network@Network{broadcast} -> do
+    withAsync (forever $ atomically (readTQueue resendQ) >>= broadcast) $ \_ ->
+      reliableBroadcast ourIndex ackCounter sentMessages network
+ where
+  allParties = fromList $ sort $ me : otherParties
+
+  reliableBroadcast ourIndex ackCounter sentMessages Network{broadcast} =
+    action $
+      Network
+        { broadcast = \msg ->
+            case msg of
+              Data{} -> do
+                ackCounter' <- atomically $ do
+                  acks <- readTVar ackCounter
+                  let newAcks = constructAcks acks ourIndex
+                  writeTVar ackCounter newAcks
+                  modifyTVar' sentMessages (insertNewMsg msg)
+                  readTVar ackCounter
+
+                traceWith tracer (BroadcastCounter ourIndex ackCounter')
+                broadcast $ ReliableMsg ackCounter' msg
+              Ping{} -> do
+                acks <- readTVarIO ackCounter
+                traceWith tracer (BroadcastPing ourIndex acks)
+                broadcast $ ReliableMsg acks msg
+        }
+
+  reliableCallback ackCounter sentMessages seenMessages resend (Authenticated (ReliableMsg acks msg) party) = do
+    if length acks /= length allParties
+      then ignoreMalformedMessages
+      else do
+        partyIndex <- findPartyIndex party
+        (shouldCallback, messageAckForParty, knownAckForParty, knownAcks) <- atomically $ do
+          let messageAckForParty = acks ! partyIndex
+          knownAcks <- readTVar ackCounter
+          let knownAckForParty = knownAcks ! partyIndex
+
+          -- handle message from party iff it's next in line OR if it's a Ping
+          --
+          -- The stream of messages from a peer is expected to look like:
+          --
+          -- @@
+          -- Msg [.., 1, ..] (Data nid m1)
+          -- Msg [.., 2, ..] (Data nid m1)
+          -- Msg [.., 3, ..] (Data nid m1)
+          -- Msg [.., 3, ..] (Ping nid)
+          -- Msg [.., 3, ..] (Ping nid)
+          -- Msg [.., 3, ..] (Ping nid)
+          -- @@
+          --
+          -- Pings are observed only for the information it provides about the
+          -- peer's view of our index
+          if messageAckForParty == knownAckForParty + 1
+            then do
+              let newAcks = constructAcks knownAcks partyIndex
+              writeTVar ackCounter newAcks
```

Comment:
This is seemingly bumping the ack counter for party to be + 1. In the code here it would be much more clear if the function would actually be named like that and/or do it here? e.g. `modifyTVar ackCounter (incAckForParty partyIndex)` or so